### ISSUE TITLE
Nav2 Integration

### ISSUE THREAD

**Issue by SteveMacenski** [Fields2Cover/Fields2Cover#73]
Hi! Thanks for the chat today! I'm happy to be working together to have out-of-the-box support for coverage planning in ROS 2 via Nav2! As discussed: 

Things we need to support:
- [ ] Start points and optionally end point setting 
- [ ] Voids in the field (Related: https://github.com/Fields2Cover/Fields2Cover/issues/38, https://github.com/Fields2Cover/Fields2Cover/tree/cellular_decomposition)
- [ ] Concave shapes (Related: https://github.com/Fields2Cover/Fields2Cover/issues/11)
- [ ] Option to have turns happen inside the polygon when leaving the polygon shape is not permissible  https://github.com/Fields2Cover/Fields2Cover/issues/76
- [ ] Overlapping operational segments throughout if op width < robot width
- [ ] have path around border option

Tentatively: 
- I'd plan to work on the ROS 2 / Nav2 integration server work + documentation / tutorials -- with some demos needed for API use for example to work from
- Vinny / Adam plan to work with Gonzalo on the technical features above 
- I think all 3 of us should get together to discuss the ROS 2 Action interface 

This will definitely make a good ROSCon talk for 2024 for Vinny/Adam/Gonzalo or a subset of them to present! 

CC @vinnnyr @Aposhian

**Comment by SteveMacenski** [comment]
@Gonzalo-Mier obviously no rush on our 3-6 months timeline, but I'm curious if or how you release this software? We'll need to be able to install it from debian binaries in the ROS build farm to release Nav2 itself in binary form. We can discuss ways to accomplish that, the ROS build farm is one possible way

**Comment by SteveMacenski** [comment]
I don't know much about cpack / pypi off hand, but those probably don't help us much. We need binaries installable via `apt` servers, which the build farm can provide for us. Pypi would be fine if we were using the python side of things, but we'll be using the C++ side. So, I don't think any of that is high priority on our account

**Comment by Aposhian** [comment]
The package is already buildable with colcon (has a package.xml) so getting it on the build farm would require:
- creating a gbp release repository
- PRing to rosdistro?
I'm not super clear on that process since I haven't gone through it myself.

**Comment by SteveMacenski** [comment]
Yeah, but another problem for another time. I'm just communicating the need and clarifying if its already available or not

**Comment by vinnnyr** [comment]
Filed:[ Option to have turns happen inside the polygon when leaving the polygon shape is not permissible ](https://github.com/Fields2Cover/Fields2Cover/issues/76)

**Comment by SteveMacenski** [comment]
^ added to check list above

**Comment by SteveMacenski** [comment]
@Gonzalo-Mier @Aposhian @vinnnyr 

I'm starting on this with a client and I had a few questions after I spent a few hours looking over the codebase, paper, and documentation. Overall, I think I have a good understanding of what I need to do and how things work, but there are a couple of conceptual or formatting things which I'm unclear on. If you could enlighten me, that would save me quite a bit of time while working to integrate this into Nav2!

- I see this phrase `Cell` and `Cells` frequently but I'm really unclear as to what this semantically means in context. I see the definition `A Cell (f2c.Cell) is a polygon created by one outter LinearRing and zero, one or many inner LinearRing.` which gives me the general jist of the kind of data it has stored, but in what situations is this data structure preferred in computations (or is it only just the highest-level structure that I can think of as just the `field`, except the `field` class contains some other metadata)?
- My understanding of this is that there is no support for "inner rings" or something with voids and non-convex shapes. Is that correct in my statement that then the `Cells` inner rings right now are just completely ignored / invalid and that this is designed in for future-proofing when it is possible?
- I'm going through the parsers trying to ID the basic types I need to be able to support for creating a ROS interface definition for the polygons and noticed that the GML parser seems to ignore [all this information](https://github.com/Fields2Cover/Fields2Cover/blob/f75aaab695275ecae389d4740d136875e01d5ece/data/envelope1.xml#L13-L54) in the test files and only gets things under the `Field->geometry->gml:Polygon->...` regime. Am I correct in saying that all we need is the ordered set of GPS waypoints? I'm surprised to see content like `area` even ignored.

**Comment by Gonzalo-Mier** [comment]
Great news, @SteveMacenski !!

Answering to your questions:
- _I see this phrase `Cell` and `Cells` frequently but I'm really unclear as to what this semantically means in context. I see the definition `A Cell (f2c.Cell) is a polygon created by one outter LinearRing and zero, one or many inner LinearRing`. which gives me the general jist of the kind of data it has stored, but in what situations is this data structure preferred in computations (or is it only just the highest-level structure that I can think of as just the `field`, except the `field` class contains some other metadata)?_

Fields2Cover inherits basic types from [GDAL and OGR](https://gdal.org/doxygen/classOGRGeometry.html). The data follows the [Simple Feature format](https://www.ogc.org/standard/sfa/), which is also used by [WKT](https://en.wikipedia.org/wiki/Well-known_text_representation_of_geometry). A `Cell`(f2c) is a [`Polygon`](https://gdal.org/doxygen/classOGRPolygon.html), and `Cells`(f2c) is a [`MultiPolygon`](https://gdal.org/doxygen/classOGRMultiPolygon.html). We decided to call it `Cell` due to the Cell decomposition algorithms family, which is used to convert a non-convex field into smaller convex cells. 

Usually, a file of a field shape contains the points of the borders in their preferred [Coordinate Reference System (CRS)](https://en.wikipedia.org/wiki/Spatial_reference_system). Some of those are not in a Cartesian plane, so working with their geometries without projecting them into another CRS is quite difficult (#87). Then, `Field` saves the metadata of the actual CRS and a reference point to the first point of the field. `Cells` is the field shape, but the first point is in `(0, 0)` (moving the field to 0 is not actual needed, but easier to work with). Remember to return the `Path` generated to the previous CRS.

To do the computations, once the coordinate are in a planar system, we can work with only `Cell` and `Cells`, without the other metadata in `Field`.

- _My understanding of this is that there is no support for "inner rings" or something with voids and non-convex shapes. Is that correct in my statement that then the `Cells` inner rings right now are just completely ignored / invalid and that this is designed in for future-proofing when it is possible?_

That's not true, but I understand the confusion. inner rings are not ignored, and, in general, algorithms on the library are prepared to handle them. When we say "no support for non-convex fields", it's more related to missing optimizers on the `route planner` and `path planner` modules and bugs related to it #11. If you are not planning to use them, it works fine: 

```
import fields2cover as f2c

rand = f2c.Random(42)
field = rand.genNonConvexField(4e3)
obstacle = rand.genNonConvexField(1e2)
field.field.addRing(0, obstacle.field.getCellBorder(0))
 
no_hl = f2c.HG_Const_gen().generateHeadlands(field.field, 3.0)
swaths = f2c.SG_BruteForce().generateBestSwaths(f2c.OBJ_NSwath(), 1, no_hl.getGeometry(0))

f2c.Visualizer.figure()
f2c.Visualizer.plot(field)
f2c.Visualizer.plot(no_hl)
f2c.Visualizer.plot(swaths)
f2c.Visualizer.show()
```
![Figure_1](https://github.com/Fields2Cover/Fields2Cover/assets/3874494/7abab346-0056-49cf-9a7f-fe870fe12dea)

- _I'm going through the parsers trying to ID the basic types I need to be able to support for creating a ROS interface definition for the polygons and noticed that the GML parser seems to ignore [all this information](https://github.com/Fields2Cover/Fields2Cover/blob/f75aaab695275ecae389d4740d136875e01d5ece/data/envelope1.xml#L13-L54) in the test files and only gets things under the `Field->geometry->gml:Polygon->...` regime. Am I correct in saying that all we need is the ordered set of GPS waypoints? I'm surprised to see content like `area` even ignored._

This file was part of a previous project. As you said, we only need the shape of the field and the CRS. `area` can be computed, once the field is in meters. The information of the `Edges` (that you mention as ignored) is redundant, and has the same info as the borders of the field (The first edge starts in the second point of the field geometry). Either way, we can agree that parsers need more work, but we need more examples to create a more generic parser :)

Thank you really much for your compliments. Feel free to ask for clarifications if needed.

**Comment by SteveMacenski** [comment]
> We decided to call it Cell due to the Cell decomposition algorithms family

Ahhhh, ok. This was not clear to me. I don't know where that could be clarified, but I'll think about that and open a PR accordingly so that the next chump like me doesn't spin in circles (you know, like a robot covering a field :wink: ). 

Thanks for the explanation. One other thing:

- Any thought as to how this could work if the coordinates aren't in GPS frame but just normal cartesian coordinates for indoor applications or applications where the Field is derived from the output of 3D or visual SLAM maps? Would it be just as simple as updating the data stored in the Cell as cartesian coordinates? Minus the polishing details. Is there anything internal tying GPS (Lat, Long) that couldn't be 1:1 replaced with X, Y?

**Comment by vinnnyr** [comment]
I don't think there is --- for context I use F2C in euclidean coordinates rather than LLA coordinates. But, I completely don't use the `Field` datatype, so maybe that's why it works?

**Comment by SteveMacenski** [comment]
One of the things I'd think about is the path planning set of things; in order to have a minimum turning radius used, there must be _some_ relationship with LLA to Cartesian coordinates, right? 

@vinnnyr do you have your own parser then? It looks like they all output `F2CFields`

**Comment by Gonzalo-Mier** [comment]
Input data in F2C is supposed to be in meters. As real field measurements
are usually done in a global CRS (long, lat), F2C provides the `Transform`
class. You can see it as an optional data pre/post-processing step.

As Vinny said, if your data is already in meters, you can forget about
`F2CField`, the `Transform` class, and just fill your `Cell` with the raw
data :)

**Comment by vinnnyr** [comment]
> Do you use all 4 of the steps (headland, swath, route, path) using cartesian?
yes, all those steps can be done in cartisean !

**Comment by SteveMacenski** [comment]
That's incredibly good to know. That won't be part of my first salvo in supporting meters, but it will be once in open source (current client doesn't need that)

**Comment by SteveMacenski** [comment]
@Gonzalo-Mier am I correct in saying that the `allow_overlap` is only applicable to the final swath being generated and not during the overall coverage planning? I believe this implies that to me https://github.com/Fields2Cover/Fields2Cover/blob/f75aaab695275ecae389d4740d136875e01d5ece/include/fields2cover/swath_generator/swath_generator_base_impl.hpp#L52 but just checking

**Comment by Gonzalo-Mier** [comment]
Yes, more information can be found in this PR (https://github.com/Fields2Cover/Fields2Cover/pull/71)

**Comment by SteveMacenski** [comment]
Great! I'm making fast work of this. Its largely going to look similar to your ROS version, but with an Action API input, lifecycle-component node and options on each step, with some cleanup (+ the nav2 integration bells and whistles with the BT nodes and what not). Honestly, I thought there would have to be more I'd have to do here. With this project, I may have bandwidth to work on some of the features listed above, but let me touch base with my client first to see if/which are of their interest.

**Comment by Gonzalo-Mier** [comment]
Cool, great news then! I'm really happy that you find the library easy to use :)

**Comment by SteveMacenski** [comment]
We'll see how I feel if/when I start modifying it :wink:

**Comment by SteveMacenski** [comment]
@Gonzalo-Mier I've spent several hours on this issue where I can't seem to get field parsed via `importGml` to work to generate headlands. It returns an empty result that causes `.getGeometry(0)` to throw `Geometry does not contain point 0`.

I've tried a number of combination of methods, including using the example files in the ROS and in this repo's `data` folder and can't seem to get it to work until I replaced the field from file with something like `auto f = rand.generateRandField(5, 1e3);` like used in the tests. Then things seem to work OK.

Are you aware of any regressions in the parsers? Are you using the GML one? Or is there something you know wrong. Examples below:

Not working: 

```
  F2CFields f;
  f2c::Parser::importGml("/home/steve/Documents/bonsai_ws/src/Fields2Cover/data/example1.xml", f);
  auto f1 = f[0].field.clone();
  f2c::hg::ConstHL hl_gen_;
  auto result = hl_gen_.generateHeadlands(f.field, 7.5);
  std::cout << result.size() << std::endl;  // ALWAYS 0
```


Working 

```
  f2c::Random rand;
  auto f = rand.generateRandField(5, 1e3);
  f2c::hg::ConstHL hl_gen_;
  auto result = hl_gen_.generateHeadlands(f.field, 7.5);
  std::cout << result.size() << std::endl;
```

I'm really stumped here. When I throw it through the swath generator without using the headland generator, it also doesn't return any results. So its not uniquely a headland generator problem. I can't get any file in this repository that's an xml to work.

Any insight / help? I'm not at a loss

**Comment by Gonzalo-Mier** [comment]
Hi again. I know it is a weird error, but I'll try to explain it

This is related to the coordinate system discussion we had before. XML
files in the data folder are in EPSG:4258 (which is not in meters).
The function 'importGml' reads the file data and writes it into a
'F2CFields' class, but it doesn't transform that data into another
coordinate system.

Then, in your non-working example, if you compute the area of 'f1', you
will obtain an area smaller than 1 (not sure of the exact result). This is
because the data is in EPSG:4258. When you compute the headland with the
width 7.5, that 7.5 is not in meters but in the EPSG:4258 coordinate
system. This is the reason why the result is always an empty Polygon.

**Comment by SteveMacenski** [comment]
Ah, OK thank you!! After another 15 minutes of playing around I've confirmed that's exactly it! Woof, thanks a million.

What do you recommend I do then w.r.t. demos and default test cases? The files are in this coordinate from that's super duper tiny (I can show after I reduce the widths to like 0.001 that it works, but also on the scale of 1e-4 on visualization hah). Should I make some new files using the transformed coordinates so that things work OK out of the box or do you have a potential other recommendation?

Also curious from your expertise what exactly is the difference between EPSG:28992 EPSG:4258. I'm just really surprised to see one's coordinates are basically 1e-4 different and another is 0.2 different. Are these really just different zones? Feels like the relative changes in distance should be preserved . . . just different offsets. But apparently I know nothing :shrug:

**Comment by Gonzalo-Mier** [comment]
No worries, we are here to help :)

Yes, new files with the field border in meters would be easier to present
without too much explanation about co-ordinate systems.

EPSG:28992 (https://epsg.io/28992) has an accuracy of less than 1 meter in
the Netherlands (only in NL) and it works in meters.
EPSG:4258 (https://epsg.io/4258) is the European version GPS coordinate
system and works in lat-long.
GPS is EPSG:4326.

There are many CRS, depending on what you are interested on. As the world
is a sphere, when working with maps you need to choose between working with
the sphere (lat-long) to preserve the aspect ratio, or to work in 2D
(meters) in a small area assuming that the results outside that area would
be horrible. Furthermore, tectonic plates moves, and those CRS are fixed to
them to avoid errors due to this movement (EPSG:4326 and EPSG:4258 offset
grows every year due to this)

**Comment by SteveMacenski** [comment]
Separately of this topic, you mentioned last time we spoke in person that a V2 might be coming out where alot of this changes or some of the Nav2 integration action items above are handled. Any update on that or desire for help (or desire for contract payment to work on a couple of those features important to particular clients sooner than later?)

**Comment by Gonzalo-Mier** [comment]
Hi @SteveMacenski,
We don't have any date for F2C V2 yet. Whenever we decide to release the V2, we will add a migration guide to make it easier for anyone, so no worries. Basic types will remain the same (except for F2CRobot and F2CPath), and some modules will have more functionalities or slightly different interface. Fortunately, mainly the complete code will be reusable :)

Unfortunately, on our side we need to comply with the embargo, so we cannot publish anything without consent.

**Comment by Gonzalo-Mier** [comment]
Sorry for closing this issue. It was a mistake

**Comment by SteveMacenski** [comment]
Higher level questions:
- If the robot / operation widths are in the GPS coordinates, that depend on zone, do most people convert the GPS into cartesian? Or are the GPS widths kind of fudged? I'll always make it an option, but I feel like converting to cartesian for this reason should be default behavior 
- For implementing "start at the nearest row" or "end at this row" behavior, do you think it would be as simple as using the snake / spiral route finder and then just renumbering them (e.g. shift) where the first is to be the one selected to start or end (but not start and end, one or the other you can control)? I was thinking about that last night and that seems like an easy enough afternoon project. That's one of the really core "out of the box quality" things missing for autonomous navigation, where its typically expected that you're taking into account the robot state in planning

**Comment by vinnnyr** [comment]
@SteveMacenski for start positions, the swath sorter has this concept of “variants”, so for a square plot, you can actually choose between variants 0-3 which correspond the all 4 corners of the square. 

https://github.com/Fields2Cover/Fields2Cover/blob/cb2b9b9963020658212bf7abbc4ea4a9f60391f6/src/fields2cover/route_planning/single_cell_swaths_order_base.cpp#L12

The remainder is determining what corner is the best to start at relative to the robot.

**Comment by SteveMacenski** [comment]
@vinnnyr what if its not a corner but starting somewhere along in the middle of an outer edge? I guess that's what I'm imagining, but the corner remark is also very good to know! Does the `N` input for `variant` translate to the `N`th node making up the outer polygon? If there's some way of knowing what that variant parameter will do, then I can expose that as an option!

**Comment by Gonzalo-Mier** [comment]
> If the robot / operation widths are in the GPS coordinates, that depend on zone, do most people convert the GPS into cartesian? Or are the GPS widths kind of fudged? I'll always make it an option, but I feel like converting to cartesian for this reason should be default behavior

Any robot measurements should be in meters. GPS coordinates are for global position, not for local measurements.

> For implementing "start at the nearest row" or "end at this row" behavior, do you think it would be as simple as using the snake / spiral route finder and then just renumbering them (e.g. shift) where the first is to be the one selected to start or end (but not start and end, one or the other you can control)? I was thinking about that last night and that seems like an easy enough afternoon project. That's one of the really core "out of the box quality" things missing for autonomous navigation, where its typically expected that you're taking into account the robot state in planning. Other things above are important, but this is one I can just feel the tickets being filed for immediately on release

As @vinnnyr mentioned, you have the variants of each pattern (More info about the 4 variants on the [tutorials](https://fields2cover.github.io/source/tutorials/route_planning.html)). For the implementation of the "start at the nearest row", just get the distance of the robot start position to the start point of the swath. Something like:

```
f2c::rp::BoustrophedonOrder swath_order;
F2CPoint start_point (0, 0);

double d, min_d {1e10};
size_t min_i;
for (size_t i = 0; i<4; ++i) {
  auto sorted_swaths = swath_order.genSortedSwaths(swaths, i);
  d = sorted_swaths[0].startPoint().Distance(start_point);
  if (d < min_d) {
    min_d = d;
    min_i = i;
  }
}
return min_i;
```

**Comment by SteveMacenski** [comment]
> Any robot measurements should be in meters. GPS coordinates are for global position, not for local measurements.

Follow up question: where does that conversion take place? 

I just carefully went through every API and their generators parameters and found a couple other hidden ones I'll want to expose as well I missed (e.g. step_angle for swath and turn_point_dist for path planner). These are pretty subtle, not sure why these aren't API method arguments. Variant I should have gotten the first time, but I think I've got them all now. 

> For the implementation of the "start at the nearest row", just get the distance of the robot start position to the start point of the swath. Something like:

I see how this blob gets is the starting row to use (which is straight  forward) - but I also want to reorder all of the swaths in the route so I can then feed that newly ordered result into the planner to get a complete solution starting at that `min_i` since there will be a wrap around somewhere if not starting in a corner

**Comment by Gonzalo-Mier** [comment]
> Follow up question: where does that conversion take place?

There is no need for a conversion. Let me try to explain myself better. To
being able to locate something in the world, we need a CRS. That CRS has
its own scale. When the scale in a region matches 1:1 with meters, we can
subtract the position of two points and obtain the vector between them. The
module of that vector is the distance. That distance is only useful for us
in that case. Luckily, we have some CRS that works fine with this
assumption. It also works fine when we are working in a local map that uses
meters.

On the other hand, if we have a robot, with values like width, length,
velocity... etc. All of that goes in international metric system, as it
doesn't depends on which continent the robot is. Then, no conversion is
required as values are expected to be provided in international metric
system.

> step_angle for swath

In the brute force algorithm, it's the difference between the angles tried.
The default value (1 degree) is picked from literature.

> turn_point_dist

In the path planner, it's the discretization distance used on the turns. A
value of 0.05 will create a path with 5 cm between each point in the
curves. If the value is 0, the default discretization is provided.

> These are pretty subtle, not sure why these aren't API method arguments.

They should. You are right.


>  I also want to reorder all of the swaths in the route

Check the custom order class in the route planner. It allows us to do that
reordering. Not sure if that's what you need.

**Comment by SteveMacenski** [comment]
>  It also works fine when we are working in a local map that uses meters. ... All of that goes in international metric system, as it doesn't depends on which continent the robot is. Then, no conversion is required as values are expected to be provided in international metric system.

Then why did I need to convert the 2.5m width to 0.0001m to make it work for the test file's default CRS (before transforming it to another)? If you scale it appropriately already, shouldn't the 2.5m or even 1.0m width been able to find a solution -- since I'm using the same GPS points, just in another CRS, both in metric? 

The widths don't seem to be being scaled :frowning:

**Comment by Gonzalo-Mier** [comment]
> Then why did I need to convert the 2.5m width to 0.0001m to make it work for the test file's default CRS (before transforming it to another)?

Because you shouldn't. To work with the test file, you need to convert it (in its CRS) to meters. Doing it in the other way is confusing and leads to many bugs. 

Following your example, 2.5m width is converted to 0.001 in the CRS metric system.
2.5 * c == 0.001, where c is a constant.
The maximum curvature for the path planner is in 1/m, then the conversion would be: k/c (where k is the max curvature).
The maximum linear curvature change is in 1/m^2, and the conversion would be: sigma/c^2 (where sigma is the max linear curv change).
This process has to be made by the user, and bugs like yours may appear.

> how can you use this code without having to "tune" your actual width measurements to the GPS CRS frame scale?

Transform the field coordinates with the Transform class:
```
int main() {
  F2CFields fields;
  f2c::Parser::importGml(std::string(DATA_PATH) + "test1.xml", fields);
  F2CField field {fields[0]};
  f2c::Transform::transform(field, "EPSG:28992");  \\ CRS in meters in NL

  ... compute path on field ...

  F2CPath returned_path = f2c::Transform::transformPathWithFieldRef(path2, field, "EPSG:4258");
}
```

**Comment by SteveMacenski** [comment]
Let me back up. I think what you just said directly conflicts what I thought you said before. 

> There is no need for a conversion. ... This process has to be made by the user

---

> Following your example, 2.5m width is converted to 0.001 in the CRS metric system.

Nothing did that for me, I manually tested 1.0m and just went down orders of magnitude of what I set the width to be until something provided me results after you mentioned the transform thing. Using [Envelope1](https://github.com/Fields2Cover/Fields2Cover/blob/main/data/envelope1.xml) from `data/` for example, its in `EPSG:4258`.

When I set something as 2.5m using that file without transforming that CRS into another CRS (e.g. `EPSG:28992`), it doesn't work. Nor does it work with 1.0m, 0.1m or 0.01m. I have to go down to 0.001m before any non-empty response is given. That tells me that the width, in meters, is not being transformed into the appropriate CRS coordinate scale and that I'm forced to do that either manually (hacky) or transforming into another `EPSG:XXXX` which simply has a different scale, of which I also do not know. As far as I can tell, there's always some scale applied of which I'm not privy to. If there isn't why doesn't the original CRS work?  

```
  f2c::Parser::importGml(std::string(DATA_PATH) + "test1.xml", fields);
  f2c::Transform::transform(field, "EPSG:28992");  \\ CRS in meters in NL
```

The data in `test1.xml` should already be in the correct CRS of the data it was taken, no? Why would it be taken in NL for example but be in a `EPSG` of another in the file? Where did the selection of `EPSG:28992` come from and why is that not already the option in the file taken the data? How should I arbitrarily know which I need to transform and which I do not? 

These are the root of my questions

**Comment by Gonzalo-Mier** [comment]
Okay, I'm happy we are in the same page now.

>  is it that the original are GPS coordinates and the EPSG:28992 is in WGS 84 cartesian coordinates?

Yes

How does one know which EPSG-prefixed number are GPS vs WGS84 cartesian?

GPS (WGS84 in America) is 4326, GNSS in Europe (ETRS89) is 4258. Examples like 28992 are particular and needs a bit of research on internet. It's usually easier to convert to UTM (`f2c::Transform::transform(field, "UTM:32N datum:WGS84")`) to work on meters.

>  because we record in GPS coordinates and later want to get these as cartesian for use in CCP?

Exactly :)

> I see nothing on these webpages that makes me think they're not just two different boxes, but actually representing two different types of data.

Attributes -> Unit -> metre / degree

> That feels like the kind of thing that should have a lookup through these standard GPS libraries, no?

I had the same thought, but I couldn't find it.

> What's the workflow to correlate them + can it be done automatically or does that need to be an attribute of the field input?

I wish it could be done automatically.

**Correction**: It CAN be done automatically. First, find the [UTM zone](https://gis.stackexchange.com/questions/209267/r-return-the-utm-zone-that-a-wgs84-point-belongs-to), and then transform it as `f2c::Transform::transform(field, "UTM:32N datum:WGS84")`. I cannot do it this week, but I'll add it next week :)

**Comment by SteveMacenski** [comment]
Thanks alot for the sounding board / helping clear up the details! 

> It's usually easier to convert to UTM (f2c::Transform::transform(field, "UTM:32N datum:WGS84")) to work on meters.

Dumb question #17: why don't the demos / tests use UTM then? I actually assumed that UTM would hook back in somewhere here that I didn't yet see, so that's why I didn't ask about that yet hah

> Correction: It CAN be done automatically. First, find the [UTM zone](https://gis.stackexchange.com/questions/209267/r-return-the-utm-zone-that-a-wgs84-point-belongs-to), and then transform it as f2c::Transform::transform(field, "UTM:32N datum:WGS84"). I cannot do it this week, but I'll add it next week :)

The `32` I imagine changes That's the thing that would change in that transform call based on the formula, correct? That would get us away from the tile lookups and such! Would a new workflow look like: 
- {I, we, you} depending on where this should live would create some method `convertToUTM()` that would auto-find the UTM zone to populate into the transform. 
- At the receiving a GPS file, call `convertToUTM()` which would handle it
- Do my thing in computing coverage
- Optionally convert back to GPS `convertFromUTM()` or use cartesian if desired

**Comment by Gonzalo-Mier** [comment]
> Dumb question https://github.com/Fields2Cover/Fields2Cover/issues/17: why don't the demos / tests use UTM then? I actually assumed that UTM would hook back in somewhere here that I didn't yet see, so that's why I didn't ask about that yet hah

It's a completely legit question. It should.

> The 32 would be the change in that transform call based on the formula, correct? That would get us away from the tile lookups and such! 

Yes, exactly. And the workflow will be as you mentioned.

> In this new next-week-world, the srsName is not super important then, since we wouldn't be using any of the EPSG:XXXX frames, yeah?

Exactly. I have added the new functions to a new branch called "feature-UTMconversion". I need to make the tests and documentation, but at least we can discuss the interface.

**Comment by Gonzalo-Mier** [comment]
@SteveMacenski  PR #95 add the easy transformation to UTM and tutorials.

**Comment by SteveMacenski** [comment]
I’ll get using that next week! Thanks!

**Comment by SteveMacenski** [comment]
@Gonzalo-Mier are you planning on releasing F2C to rosdistro for binaries? While not incredibly time sensitive, it would be necessary before this server can be released into Nav2 for its own binaries since we can't use git submodules in the build farm. My plan today is to finish up the Nav2 server + tests + UTM and start on the integration work with Behavior Trees later in the week or next week.

**Comment by Gonzalo-Mier** [comment]
> are you planning on releasing F2C to rosdistro for binaries?

Not sure how to do that. If you have a tutorial, I could do it :)

**Comment by SteveMacenski** [comment]
An update: I should be ready to move the repo to a public setting sometime next week. I want to  talk with the client first and make sure they're happy before I start talking about it publicly

@vinnnyr @Aposhian what's the timeline / plan with respect to some of the F2C improvements for your application described above? I think I've contributed my part of this arrangement :wink: I recall the plan was Jan 2024, but just wanted to make sure this was still on the mind / on track for early next year. I'll need to speak with my client, but I may be able to help with a couple of them, but that's TBD. 

@Gonzalo-Mier thanks for all your help and hopefully you approve of my final work and the Nav2 / ROS 2 integration. There are videos!

**Comment by SteveMacenski** [comment]
@Gonzalo-Mier any plans for binary release? That's the only thing blocking me from being able to release the newly released packages into binaries: https://github.com/open-navigation/opennav_coverage

**Comment by Gonzalo-Mier** [comment]
Hi @SteveMacenski, I've just released it. Can you check it if it is correct?

**Comment by SteveMacenski** [comment]
Next thing is to bloom it to the ROS build farm so that it can be available via binaries. Here's my guide: https://gist.github.com/SteveMacenski/1c321d1c9edca096ae4d763d8327c2ee. You already have a package.xml so it should be easy

**Comment by Gonzalo-Mier** [comment]
Thanks for your support!
PR created: https://github.com/ros/rosdistro/pull/38990, https://github.com/ros/rosdistro/pull/38991, https://github.com/ros/rosdistro/pull/38992

**Comment by SteveMacenski** [comment]
Rolling too? That should button it up!

**Comment by Gonzalo-Mier** [comment]
Done! https://github.com/ros/rosdistro/pull/38999

**Comment by SteveMacenski** [comment]
Great! On the next rolling/humble sync, I can remove the manual F2C builds from our CI & check that item off of our Nav2 migration list! Also I'm sure many of your users will thank you :- )

**Comment by Gonzalo-Mier** [comment]
Thank you for your support!

**Comment by vinnnyr** [comment]
@SteveMacenski  for the current open nav coverage planner, can we just handle start / end point with having a BT call Nav2 `NavigateToPose` for the start point of the plan?

**Comment by SteveMacenski** [comment]
That would be pretty easy to bake into the BT itself, but I was thinking more about having the actual `route` stage consider the row who's start is closest to the robot's current position should be set as start, so that we maximize efficiency. On the other side of the coin, having the row's end that is closest to a potential "goal" be the end of the `route`, to minimize post-coverage navigation. I don't think both are wise to optimize for simultaneously, but they involve the same operation, just inverted (and in fact, could be computed the same, where the orderings are just inverted). 

Perhaps not the most core of the core requirements above.

**Comment by SteveMacenski** [comment]
@Gonzalo-Mier I'm working on a new feature which I can highlight at a later time and running into some transformation issues. Path is the only function with a `transformXWithFieldRef` method, the others have only `transformX`. Is there a reason why strip(s) and swath(s) don't have that option and used in the `transformToPrevCRS` implementations?

**Comment by SteveMacenski** [comment]
Got it! We ended up doing something slightly different. I may still yet contribute those transform functions, but we ended up implementing the transforms on our own Swath-like type before converting them into swaths simply due to this particular program's ergonomics. 

BTW: this feature is allowing the use of F2C with a pre-computed set of rows based on existing information and/or physical infrastructure dictating the position and attributes of each swath. It should be available before EOY 

Oh wow, awesome! I think there's still a few things missing, but to me that pushes it over the line as 'enough' to include in Nav2 more directly. How bad are the API changes? :fearful:

**Comment by Gonzalo-Mier** [comment]
I would say not too bad. The F2CRobot class is the one that will change more. The other changes will be renaming a few functions and moving public variables in classes to protected/private.

I'll make a document on how to make those changes, to avoid future headaches.

**Comment by SteveMacenski** [comment]
Got it. I was curious if you spent any time thinking about articulation for the robot class? Even if not for self-collisions, do you think the current linear curvature change limit is sufficient to consider that? I'm thinking if you play it very conservatively, due to the structured nature of the coverage task, I think it is OK since you've always straightening out after the curve to go down the row and then also starting in a straight configuration before a turn.

**Comment by Gonzalo-Mier** [comment]
> I was curious if you spent any time thinking about articulation for the robot class?
We thought about it, but we have so many different problems to solve that this was not one of our priorities. The new robot class has more options to define the shape of the robot, and zero or one fixed implement attached. If needed, we can add the possibility to attach other implement to an implement, so they can be chain. Right now we were only considering fixed implements, but we are open to any progress in this sense.

> PS: I should have my updates available publicly in about a week! It allows for use of F2C with non-autogenerated swaths. So you can use preannotated swaths from a physical field and then use the rest of F2C as if it were normal. There's also a skipping feature to skip annotated rows for whatever reason in the request field
Really interesting!

**Comment by Gonzalo-Mier** [comment]
> Great! On the next rolling/humble sync, I can remove the manual F2C builds from our CI & check that item off of our Nav2 migration list! Also I'm sure many of your users will thank you :- )

@SteveMacenski, since today F2C can be installed with `sudo apt install ros-iron-fields2cover` :partying_face: Waiting for the others to be updated

**Comment by SteveMacenski** [comment]
Any thoughts as to by `Fields2Cover` can't be found? The build farm job to release the ROS action server failed on [find_package](https://github.com/open-navigation/opennav_coverage/blob/iron/opennav_coverage/CMakeLists.txt#L18) for it. It didn't complain about the [rosdep key](https://github.com/open-navigation/opennav_coverage/blob/iron/opennav_coverage/package.xml#L21), so it found the package and I see in the job that it installed it properly. Your here CMakeLists is a little different than I'm used to scanning through to find potential issues. Do you know if anyone else is successfully using your binaries to build?

https://build.ros2.org/job/Idev__opennav_coverage__ubuntu_jammy_amd64/1/

**Comment by SteveMacenski** [comment]
Ah, I'd have to spin up an entire new release to update the failing build farm job, so it probably just makes sense to wait until you get a chance to fix it since I'd run a release, then you'd fix it, and then I'd have to run a new release to remove it

**Comment by mikolodz** [comment]
> Btw, the version 2 release is approved for end of January. It will contain:
> 
> * Start points and optionally end point setting
> * Voids in the field (Related: [How to deal with internal static obstacle？ #38](https://github.com/Fields2Cover/Fields2Cover/issues/38), https://github.com/Fields2Cover/Fields2Cover/tree/cellular_decomposition)
> * Concave shapes (Related: [Support more complicated shapes #11](https://github.com/Fields2Cover/Fields2Cover/issues/11))
> * Overlapping operational segments throughout if op width < robot width

Hello @Gonzalo-Mier!
I'm really excited about your v2 release! It will be a game changer for all of these smaller projects and startups.

I read through this issue, so I realize you're short on time, although could you please confirm if it will contain the features that you listed couple of moths ago and if it's still planned to be released this or next month? I ask just to make sure if we should wait until the release before we start working on our fork. Thank you for this awesome project!

**Comment by Gonzalo-Mier** [comment]
Hi @mikolodz, 
Thanks for the nice words. The release is approved, but as you have read it, I don't have that much time this month. High probably to be released at the beginning of next month.

I can share an example of a complete route: 
![complete_route](https://github.com/Fields2Cover/Fields2Cover/assets/3874494/20402055-f1c0-4562-976c-18c24a77e504)
 
Thanks for your waiting

**Comment by mikolodz** [comment]
Thank you so much! I'm happy to see such progress of your project and hear that it's coming soon. It's a wonderful news for the whole ROS2 community!

**Comment by SteveMacenski** [comment]
@Gonzalo-Mier any update :smile:

**Comment by tonynajjar** [comment]
I'll add myself to the list of people eagerly waiting for V2 :rocket:

**Comment by SteveMacenski** [comment]
I would love your help in getting those features exposed into the ROS 2 wrapper! That + the F2C binaries not working in the build farm due to some issues (@Gonzalo-Mier ETA on that fix released for Iron?) and we can move that out of `open-navigation/` into `ros-planning/` or `nav2` itself

**Comment by Gonzalo-Mier** [comment]
Hi @SteveMacenski and @tonynajjar,

Thank you for your patience! V2 still needs some work, but mainly on documentation and small checks. I plan to finish it at the beginning of next week, but meanwhile I have uploaded the code to the [v2.0.0-dev branch](https://github.com/Fields2Cover/Fields2Cover/tree/v2.0.0-dev). With this, I hope you can start working on it.

Main changes for the migration are described [here](https://github.com/Fields2Cover/Fields2Cover/blob/v2.0.0-dev/migration_to_v2.txt)

**Comment by SteveMacenski** [comment]
@Gonzalo-Mier did you ever re-release with the fix for the colcon build, build farm issue for Iron? I still see those jobs failing daily for `opennav_coverage*` due to F2C issue you mentioned fixed

**Comment by Gonzalo-Mier** [comment]
Hi @SteveMacenski,

Yes, I re-released but it had two new errors:
  1. It cannot find tinyxml2 (which is built by the project)
  2. libgdal-dev cannot be installed by apt
Each of them affect a different distribution.

On the other hand, v2.0.0 is now released and documentation is updated. I'm working on releasing to ROS, but I need to create a ros library for or-tools to be able to compile. Not sure if you know an easier way of doing this.

I've checked those tasks that are already done on the plan :)

**Comment by Aposhian** [comment]
For now, would it make sense to just pull in or-tools with `FetchContent` and build from source, or maybe have that as a build option that gets enabled for the ros buildfarm?

**Comment by SteveMacenski** [comment]
> It cannot find tinyxml2 (which is built by the project)

Why not use the tinyxml2 binaries? I'm pretty sure you can apt install this. 

> libgdal-dev cannot be installed by apt ... on [some] distribution

Which does this affect? 

> On the other hand, v2.0.0 is now released and documentation is updated. I'm working on releasing to ROS, but I need to create a ros library for or-tools to be able to compile. Not sure if you know an easier way of doing this.

Any migration guide, list of new features, etc? I can't seem to find it :smile: I'd love to know what new widgets exist to use! 

While I don't have a client project right now to justify too much of my time on it (though maybe its not a huge push?), hopefully someone in the community that wants to use the new features with Nav2 can contribute them to `opennav_coverage` or if you're open to adding those features instead of a brand new project, that could work too!   

Typically how you can wrap and release something for the ROS build farm is using vendorization. For examples:
- https://github.com/SteveMacenski/spatio_temporal_voxel_layer/tree/ros2/openvdb_vendor (+ discussion https://github.com/SteveMacenski/spatio_temporal_voxel_layer/issues/232#issuecomment-1112111566 since this is recent)
- https://github.com/ros2/yaml_cpp_vendor -- more standard use-case

These basically just fetch content and then build, but in the build farm, for packages that don't have their own released binaries. Then you make your project dependent on the vendor, NOT the core package itself, to link against this vendorized solution rather than another-possible-installed version. 

I suppose, and this is totally up to you, another option is that you fetch content so that F2C can build and instead we add a `f2c_vendor` in `opennav_coverage` to build F2C + or-tools and so on and have the ROS wrapper release F2C into the ROS ecosystem. I would prefer not, since that removes it out of your direct control, but I suppose an option depending on how you think about vendorization scope.

**Comment by ar0ser0** [comment]
> Any migration guide, list of new features, etc?

These are the most useful resources that I have found, other than digging through the source itself.

https://github.com/Fields2Cover/Fields2Cover/blob/main/migration_to_v2.txt

https://fields2cover.github.io/index.html

**Comment by Gonzalo-Mier** [comment]
>  Why not use the tinyxml2 binaries? I'm pretty sure you can apt install this.

That makes sense

> Which does this affect?

Rolling (noble), but it is now solved.

> Any migration guide, list of new features, etc? I can't seem to find it 😄 I'd love to know what new widgets exist to use!

Sorry for the delay. Here you have it: 

https://fields2cover.github.io/index.html#f2c-version-2-0
https://github.com/Fields2Cover/Fields2Cover?tab=readme-ov-file#f2c-version-20
https://fields2cover.github.io/source/migration_to_v2.html


There are a few new undocumented functions (https://github.com/Fields2Cover/Fields2Cover/blob/75ab663da4dc67966632f73908dc771567426a56/include/fields2cover.h#L151). Unfortunately, I didn't have time to test them properly.

> These basically just fetch content and then build, but in the build farm, for packages that don't have their own released binaries. Then you make your project dependent on the vendor, NOT the core package itself, to link against this vendorized solution rather than another-possible-installed version. There's a nice ament_vendor util for this purpose which wraps fetch content and version controls things

I tried both solutions but I couldn't make them work. For now, downloading the ortools binaries works, but I'm not 100% convinced as they are OS dependant.






https://github.com/Fields2Cover/Fields2Cover?tab=readme-ov-file#f2c-version-20

**Comment by Gonzalo-Mier** [comment]
I was wrong about or-tools from binaries. ROS buildfarm does not like this solution. I have tried doing it from source but it doesn't worked either. I have to put more effort on it I guess.

**Comment by SteveMacenski** [comment]
Vendorization? 

It didn’t work with fetch content? That should work (essentially what ament vendor does with more details)

**Comment by ikcalB** [comment]
@jkoubs have you tried integrating FC v2 with nav2?

Or has anyone done previous work on this?
If not, we might be able to contribute on this.

Any pointers on where to start (except the migration guide)?
